src/libsyntax/parse/parser.rs

```diff
@@ -1777,7 +1777,25 @@ impl<'a> Parser<'a> {
                    require_name);
             let pat = self.parse_pat()?;
 
-            self.expect(&token::Colon)?;
+            if let Err(mut err) = self.expect(&token::Colon) {
+                // If we find a pattern followed by an identifier, it could be an (incorrect)
+                // C-style parameter declaration.
+                match self.parse_ident() {
+                    Ok(ident) => {
+                        let span = pat.span.with_hi(ident.span.hi());
+                        err.span_suggestion_with_applicability(
+                            span,
+                            "declare the type after the parameter binding",
+                            format!("{}: {}", ident, pprust::pat_to_string(&pat)),
```

Comment:
Instead of `pprust::pat_to_string` try using `span_to_snippet` and only use `pat_to_string` if it's not available. We want to use the users' code whenever it is possible as is (think if they had written `&    Bar`, but the pattern is read produced as `&foo::Bar`).